feat: MultiStartGradient value/gradient NaN step diagnostics

## Overview

`MultiStartGradient` detects a dead lane only via the **value** (`alive = np.isfinite(np.asarray(foms))`). The gradient is never checked inside the fit loop, so a lane whose likelihood is finite but whose gradient is non-finite is counted alive, has its update zeroed by `optax.apply_if_finite`, and silently freezes in place. That failure mode — a differentiability failure rather than an evaluation failure — is currently invisible.

This adds per-step, disjoint accounting for both: **value-NaN** lane-steps (likelihood undefined) and **gradient-NaN** lane-steps (likelihood defined but not differentiable), persisted into `search_internal` and `samples_info` and surfaced in `search.summary` as raw counts plus rates normalised by `n_starts × total_steps`.

Scope is **measurement only** — `resurrect` behaviour is unchanged, so existing benchmark numbers stay comparable. A standalone demonstration in `autolens_profiling` shows the accounting does not cost runtime.

## Plan

- Add two per-step lane-death counters to the `MultiStartGradient` fit loop: value-NaN and gradient-NaN lane-steps.
- Keep the counters **disjoint** — gradient-NaN counts only lanes still alive by value, so a lane non-finite in both is one value-NaN, not one of each.
- Compute gradient finiteness **inside the existing jitted call** (returned as a third output), so there is no extra kernel dispatch and no extra host round-trip.
- Count unconditionally, not under `if self.resurrect` — with `resurrect=False` the value-NaN counter is the only record that lanes died.
- Persist both into `search_internal` (resume-safe) and `samples_info`, alongside `n_resurrections`.
- Emit a guarded block in `search.summary` with raw counts and normalised rates, following the existing MCMC duck-typed precedent.
- Extract the per-step accounting into a pure-NumPy helper so it is unit-testable without JAX.
- Add a standalone overhead demonstration to `autolens_profiling`, with a duplicate-baseline control so the measurement noise floor is reported alongside the overhead.

<details>
<summary>Detailed implementation plan</summary>

### Work Classification

**Library (PyAutoFit) + profiling artifact (autolens_profiling)** — library-led; both repos share one worktree and one branch. Two PRs; the profiling PR merges second, since its numbers describe the merged code.

### Affected Repositories

- `PyAutoFit` (primary) — the feature
- `autolens_profiling` — the standalone runtime demonstration

### Branch Survey

| Repository | Current Branch | Dirty? |
|-----------|---------------|--------|
| ./PyAutoFit | main | clean |
| ./autolens_profiling | main | clean (untracked `results/` artifacts only) |

**Suggested branch:** `feature/multistart-nan-step-diagnostics`

**Worktree root:** `~/Code/PyAutoLabs-wt/multistart-nan-step-diagnostics/`

`worktree_check_conflict multistart-nan-step-diagnostics PyAutoFit autolens_profiling` returns clean (exit 0).

⚠️ **Two unregistered worktrees on `autolens_profiling`**, neither in `active.md`:
- `feature/numerical-hazard-profiling` — 0 commits ahead of main (fully merged); stale leftover, safe to ignore, worth a separate cleanup.
- `feature/point-source-defaults-campaign` — 8 commits ahead; genuinely unmerged work, but scoped to point-source defaults and touching nothing here.

Neither is a conflict; recorded so a later session does not rediscover them.

### Implementation Steps

**1. `autofit/non_linear/search/mle/multi_start_gradient/search.py` — the counters**

- Add a pure-NumPy static helper `_nan_lane_counts(foms, grad_finite) -> (n_value_nan, n_grad_nan)`:
  - `alive = np.isfinite(np.asarray(foms))`
  - `n_value_nan = int((~alive).sum())`
  - `n_grad_nan = int((alive & ~grad_finite).sum())` — disjoint by construction
  - Returns `alive` too, so the existing resurrection path reuses it rather than recomputing.
- Extend `batched_value_and_grad` (~line 495-530) to return a third output, `jnp.all(jnp.isfinite(grads), axis=1)`, an `(n_starts,)` bool:
  - unbatched path: fold into the `jax.jit(jax.vmap(...))` wrapper;
  - chunked path: concatenate the third output per chunk alongside `foms`/`grads`, discarding padded rows the same way.
- In the fit loop (~line 625), consume the third output and accumulate `n_value_nan_lane_steps` / `n_grad_nan_lane_steps`. Counted **unconditionally**, outside the `if self.resurrect` guard.
- Initialise both to `0` on the fresh path (~line 574); restore via `int(search_internal.get(..., 0))` on the resume path (~line 549), matching the `n_resurrections` precedent.
- Add both keys to the `search_internal` dict (~line 700).

**2. `search.py` — `samples_via_internal_from` (~line 935)**

Add both counters to `samples_info` via `.get(..., 0)` so pre-existing `search_internal` files load without a KeyError.

**3. `autofit/text/text_util.py` — `search_summary_from_samples` (line 115)**

Add a guarded block following the `hasattr(samples, "total_accepted_samples")` idiom three lines above, keyed off `getattr(samples, "samples_info", {})` with `.get()`:

```
Resurrections = 797
Value-NaN Lane-Steps = 797
Gradient-NaN Lane-Steps = 12
Value-NaN Lane-Step Rate = 0.0166
Gradient-NaN Lane-Step Rate = 0.00025
```

Rates divide by `n_starts * total_steps`, guarded against a zero denominator. Raw counts are not comparable across runs (797 on a 16x3000 run vs 10 on an 8x300 run differ 80x raw but ~2x by rate).

**Naming constraint:** neutral factual counts only. Do **not** label these a smoothness metric in user-facing output — the resurrection-rate to HMC-divergence-rate correlation is unvalidated (separate item, wsdev#117).

**4. Tests**

`test_autofit/non_linear/search/mle/test_multi_start_gradient.py` is deliberately JAX-free (see its own comment: `_fit` needs jax + optax). The helper extraction in step 1 is what keeps it that way.

- `_nan_lane_counts` tested directly with NumPy arrays: all-finite; value-NaN only; grad-NaN only; both non-finite on the same lane (asserts disjointness).
- `samples_info` plumbing tested by hand-building a `search_internal` dict — the pattern already used by `test__samples_via_internal_from` and `test__samples_info__stop_reason_max_steps_and_legacy_search_internal` — including a legacy dict with neither key, asserting `0`.
- `test_autofit/text/test_text_util.py`: summary block with a stub samples object, plus a negative test that a samples object without the keys emits no block.

**5. `autolens_profiling` — standalone overhead demonstration**

`scripts/misc/searches/multi_start_nan_accounting_overhead.py`

The feature is unconditional, so there is no flag to A/B inside one process. Instead the script measures the two halves separately and divides:

1. **Denominator** — time `jax.jit(jax.vmap(jax.value_and_grad(...)))` on a real PyAutoLens likelihood at realistic `(n_starts, n_params)` shapes. That is the loop's per-step cost, same code path, same shapes.
2. **Numerator** — the accounting variants (host pull / eager device reduce / fused-in-jit) applied to that same real gradient output.
3. **Control** — a duplicate baseline run, so the noise floor prints alongside the overhead.

The control is required, not decorative: without it the script can only report "we measured no impact", which is unfalsifiable. With it, it reports "overhead is below a noise floor of X%", a claim that could have failed.

Conventions: run from repo root (`ruff.toml` sentinel walk-up); honour `AUTOLENS_PROFILING_SMOKE=1` at module top; write a version-stamped JSON + PNG pair under `results/searches/multi_start_nan_accounting/` per `results/README.md`; `ruff check` + `ruff format --check` clean (the PR gate); short pointer added to `scripts/misc/searches/README.md`. Add to the hand-maintained 5-script smoke list in `.github/workflows/lint.yml`.

Cross-reference `scripts/misc/hazards/checks/nonfinite_gradient.py` — an existing *detector* for non-finite gradients on likelihood surfaces. Related motivation, different artifact; not a duplicate.

### Prior measurement

A scratch microbenchmark (CPU, `n_starts=16`, `n_params=30`, best-of-7, duplicate-baseline control) already established the shape of the answer and drove the design:

| variant | cheap obj (180 us/step) | expensive obj (1.9 ms/step) |
|---|---|---|
| *noise floor* | ~1% | ~1.5% |
| host pull of `grads` | +6.4% | +0.9% (below noise) |
| eager `jnp.all(...)` outside the jit | +22.6% | +3.3% |
| fused into the jitted call | +5.2% | +0.05% (below noise) |

The eager-outside-the-jit variant is the **worst** of the three — it pays an un-jitted kernel dispatch *plus* a second host round-trip — which is why the plan fuses into the jitted call instead. Note the CPU benchmark structurally understates the host-pull variant, since on CPU "device to host" is a memcpy in the same address space; on GPU it is a real round-trip. The fused variant sidesteps the question on either backend.

The `autolens_profiling` script supersedes this scratch run with real likelihood shapes, and should also be run under the GPU profile.

### Key Files

- `autofit/non_linear/search/mle/multi_start_gradient/search.py` — fit loop, `batched_value_and_grad`, `search_internal`, `samples_via_internal_from`
- `autofit/text/text_util.py` — `search_summary_from_samples` (line 115)
- `test_autofit/non_linear/search/mle/test_multi_start_gradient.py` — JAX-free unit tests
- `test_autofit/text/test_text_util.py` — summary block tests
- `autolens_profiling/scripts/misc/searches/multi_start_nan_accounting_overhead.py` — the demonstration

### Deliberately out of scope

Making `resurrect` trigger on non-finite gradients. That would change search behaviour and shift every existing benchmark number, so the wsdev #117/#125 pixelized results would stop being comparable without re-running. The resurrection policy is decided **after** the counters show how often frozen lanes actually occur.

</details>

## Original Prompt

<details>
<summary>Click to expand starting prompt</summary>

```markdown
# MultiStartGradient value-NaN and gradient-NaN step diagnostics

Type: feature
Target: PyAutoFit
Repos:
- PyAutoFit
Difficulty: small
Autonomy: supervised
Priority: normal
Status: formalised

MultiStartGradient value-NaN and gradient-NaN step diagnostics.

Add per-step non-finite accounting to MultiStartGradient and surface it. Scope is MEASUREMENT ONLY — no change to resurrection behaviour (see 'Deliberately out of scope' below).

PART 1 (the key diagnostic) — split and record value-NaN vs gradient-NaN steps.
In autofit/non_linear/search/mle/multi_start_gradient/search.py the fit loop detects a dead lane ONLY via 'alive = np.isfinite(np.asarray(foms))' (~line 627) — the VALUE. The gradient is never checked inside the loop; gradient finiteness is tested only when drawing initial starts (~line 876). So a lane whose value is finite but whose gradient is non-finite is not counted dead and not resurrected: optax.apply_if_finite zeroes its update and the lane silently freezes in place while still counted alive. That failure mode is currently invisible, and it is exactly a differentiability failure rather than an evaluation failure. Count both, per step, and record them separately:
  - value-NaN lane-steps: where the likelihood is UNDEFINED (today's resurrection trigger)
  - gradient-NaN lane-steps: where the likelihood is defined but NOT DIFFERENTIABLE (new; currently unmeasured)
'grads' is already in hand at that point in the loop. Persist both counters into search_internal alongside n_resurrections, and into samples_info in samples_via_internal_from. Measure, do not assume, the cost: pulling grads to host is a larger device-to-host transfer than foms (n_starts x n_params vs n_starts) and sits inside the stepped loop — the step already syncs on foms, so it is likely in the noise, but benchmark it.

PART 2 — surface the counters in search.summary.
autofit/text/text_util.py:115 search_summary_from_samples(samples) already receives the samples object, and samples_via_internal_from already puts n_resurrections, n_starts, n_steps, total_steps and resurrect into samples_info. Add a guarded block emitting the resurrection count, the two NaN counters, and NORMALIZED rates (divide by n_starts * total_steps — raw counts are not comparable across runs: 797 on a 16x3000 run vs 10 on an 8x300 run differ 80x raw and ~2x by rate). Follow the existing duck-typed precedent three lines above, 'if hasattr(samples, total_accepted_samples)', which adds Total Accepted Samples / Acceptance Ratio for MCMC searches — search-specific blocks are already the idiom. Guard with .get() so other searches are unaffected.

Naming constraint: emit these as neutral factual counts. Do NOT label them a smoothness metric in user-facing output — the resurrection-rate to HMC-divergence-rate correlation is unvalidated (that validation is a separate ideas.md item, wsdev#117 resurrection diagnostics).

Deliberately out of scope: making resurrect trigger on non-finite gradients. That would change search behaviour and shift every existing benchmark number, so the wsdev #117/#125 pix results would stop being comparable without re-running. Decide the resurrection policy AFTER the counters show how often frozen lanes actually occur.

Motivation: pixelized-mesh MultiStartProdigy campaigns (wsdev #117/#125) judge mesh differentiability indirectly, from final logL and raw resurrection counts. The value/gradient NaN split directly answers the open question in pix_prodigy_laptop_gpu_findings.md section 6.2 — whether DelaunayNN's 109 free-AdaptSplit lane deaths were NaN deaths like plain Delaunay or survivable over-regularized-floor deaths like knn — and would have caught frozen zombie lanes in every run to date, some of which may have been misattributed to regularization plateaus.

Sizing note: this is SMALL despite the prose length — roughly a counter plus two dict keys in search.py, and a ~5-line guarded block in text_util.py, plus a benchmark and unit tests.

<!-- formalised by the Intake (Conception) Agent on 2026-08-14 from user-intake -->
```

</details>

## Library PR Created

- **PyAutoFit:** https://github.com/PyAutoLabs/PyAutoFit/pull/1473 (`pending-release`)
- **autolens_profiling:** https://github.com/PyAutoLabs/autolens_profiling/pull/127 (`pending-release`) — the overhead demonstration; merge **after** #1473, since its numbers describe the merged library code.

### API impact

`None — internal changes only.` Two additive `samples_info` keys and gradient-search-only `search.summary` lines. `grep` across `autofit_workspace`, `autogalaxy_workspace`, `autolens_workspace`, `autolens_workspace_test` and `HowToLens` for the changed surface returns zero references, so no workspace migration is owed.

### Smoke tests — all six workspaces

Run via `pyauto-heart smoke --root <task worktree>`, so the branch's PyAutoFit was the imported source (confirmed: Heart's runner replaces `PYTHONPATH` from `--root` and preflights that every module resolves under it).

| workspace | result |
|---|---|
| `autofit` | pass |
| `autogalaxy` | pass |
| `autolens` | pass |
| `autolens_test` | 3 failed (pre-existing — see below) |
| `euclid` | 6/6 passed |
| `HowToLens` | 50 passed, 0 failed |

<details>
<summary>The three failures are pre-existing on <code>main</code></summary>

```
interferometer/jax_likelihood/rectangular.py
interferometer/jax_likelihood/mge.py
multi_dataset/jax_likelihood/mge.py
```

All are stale pinned-value assertions, marginally over tolerance:

```
AssertionError: Not equal to tolerance rtol=0.0001, atol=0
 [0]: -3163.8939270532364 (ACTUAL), -3164.286252 (DESIRED)
Max relative difference among violations: 0.00012399
```

**Control run:** the same workspace, same harness, same cached environment, with `--root` pointed at canonical `main` produced the **same three failures with byte-identical ACTUAL and DESIRED values**. They are not caused by this branch — and independently, the failing scripts contain no `MultiStart` reference at all, so they never execute the changed code.

Filed as `PyAutoMind/draft/bug/workspaces/jax_likelihood_pins_stale_by_1e4.md`.

</details>

### Verification beyond the test suite

- Device-side `jnp.all(jnp.isfinite(grad))` cross-checked against host recomputation on real gradients straddling a non-differentiability cliff — exact agreement.
- Gradient-NaN counter proven to **fire**: forced a value-finite/gradient-NaN lane via the `jnp.where` AD trap → 14/80 lane-steps counted, `n_resurrections` unchanged (frozen lanes counted, deliberately not resurrected).
- `search.summary` verified **on disk**, with rates checked against `count / (n_starts × total_steps)` read off the file.
- Both guards mutation-tested — removing either fails tests, including the pre-existing `test__search_summary_to_file`.

### Known gap

Resume accumulation is unit-test-only. Demonstrating it end-to-end requires resuming a killed mid-run search, which fails on `main` for unrelated reasons (FoM sanity check compares a stored log-likelihood against the multi-start chi-squared convention). Filed as `PyAutoMind/draft/bug/autofit/multistart_gradient_resume_fom_sanity_check.md`.

Heart was **YELLOW** at ship time (score 70, `red_reasons: []`); both reasons — pre-existing autogalaxy notebook validation failures and tenant-firewall manifest drift — are unrelated to this branch and were acknowledged by the human before shipping.

--- a/autofit/non_linear/search/mle/multi_start_gradient/search.py
+++ b/autofit/non_linear/search/mle/multi_start_gradient/search.py
@@ -216,6 +216,45 @@ class AbstractMultiStartGradient(AbstractMLE):
         return bool(converged) or total_steps >= self.n_steps
 
     @staticmethod
+    def _nan_lane_counts(foms, grad_finite):
+        """
+        Split one step's non-finite lanes into the two failure modes, and return
+        the alive mask the caller needs anyway.
+
+        Two distinct things go wrong per lane, and conflating them hides the
+        second:
+
+        - **value-NaN** — the likelihood is *undefined* at this point. This is
+          the lane death ``resurrect`` already triggers on.
+        - **gradient-NaN** — the likelihood is defined but *not differentiable*.
+          Nothing detects this today: ``optax.apply_if_finite`` zeroes the
+          update, so the lane silently freezes in place while still counting as
+          alive. A frozen lane looks exactly like a converged one in the
+          figure-of-merit trace.
+
+        The two counts are **disjoint by construction**: a lane that is
+        non-finite in both is one value-NaN, not one of each. A dead lane's
+        gradient is almost always non-finite too, so counting them independently
+        would double-report the ordinary case and drown the interesting one.
+
+        Returns ``(alive, n_value_nan, n_grad_nan)``. ``alive`` is returned
+        rather than recomputed by the caller so the finiteness test on ``foms``
+        happens once per step.
+
+        Pure NumPy and free of search state so it can be tested directly;
+        ``_fit`` itself needs jax + optax + a JAX-traceable ``Analysis`` and
+        cannot be driven from the NumPy-only library suite (same reasoning as
+        ``_is_final_boundary`` above).
+        """
+        alive = np.isfinite(np.asarray(foms))
+        grad_finite = np.asarray(grad_finite).astype(bool)
+
+        n_value_nan = int(np.count_nonzero(~alive))
+        n_grad_nan = int(np.count_nonzero(alive & ~grad_finite))
+
+        return alive, n_value_nan, n_grad_nan
+
+    @staticmethod
     def _stop_reason_on_resume(stop_reason):
         """
         The ``stop_reason`` a resumed run should start from, given the one
@@ -503,7 +542,26 @@ class AbstractMultiStartGradient(AbstractMLE):
         # numerically identical to the vmap; `batch_size` never changes
         # results, it only bounds memory and compile.
         _value_and_grad = jax.value_and_grad(fitness.call)
-        _vmapped = jax.jit(jax.vmap(_value_and_grad))
+
+        # Per-start gradient finiteness is reduced *inside* the jitted call, as a
+        # third output, rather than by a separate op on its result. The reduction
+        # itself is trivially cheap either way; what costs is dispatch. Fused,
+        # XLA folds the ``isfinite``/``all`` into the backward pass and the step
+        # returns one extra ``(n_starts,)`` bool alongside the foms it already
+        # syncs on — no extra kernel launch, no extra device->host round-trip.
+        # Reducing eagerly outside the jit instead measured ~+3% per step against
+        # a ~1.5% noise floor (worse than pulling the whole ``(n_starts, ndim)``
+        # gradient to host, which is what it was meant to avoid); fused it is
+        # +0.05%, below noise. See
+        # ``autolens_profiling/scripts/misc/searches/multi_start_nan_accounting_overhead.py``.
+        #
+        # ``_value_and_grad`` is deliberately left as a 2-tuple: the single-point
+        # initial-draw path below jits it directly and unpacks two values.
+        def _value_and_grad_finite(vector):
+            fom, grad = _value_and_grad(vector)
+            return fom, grad, jnp.all(jnp.isfinite(grad))
+
+        _vmapped = jax.jit(jax.vmap(_value_and_grad_finite))
 
         if self.batch_size is None:
             self._warn_if_unbatched_exceeds_memory(model=model, analysis=analysis)
@@ -514,21 +572,25 @@ class AbstractMultiStartGradient(AbstractMLE):
             def batched_value_and_grad(params):
                 foms_chunks = []
                 grads_chunks = []
+                grad_finite_chunks = []
                 for lo, hi, pad in _chunk_slices(params.shape[0], batch_size):
                     chunk = params[lo:hi]
                     if pad:
                         chunk = jnp.concatenate(
                             [chunk, jnp.tile(chunk[-1:], (pad, 1))]
                         )
-                    foms, grads = _vmapped(chunk)
+                    foms, grads, grad_finite = _vmapped(chunk)
                     if pad:
                         foms = foms[:-pad]
                         grads = grads[:-pad]
+                        grad_finite = grad_finite[:-pad]
                     foms_chunks.append(foms)
                     grads_chunks.append(grads)
+                    grad_finite_chunks.append(grad_finite)
                 return (
                     jnp.concatenate(foms_chunks),
                     jnp.concatenate(grads_chunks),
+                    jnp.concatenate(grad_finite_chunks),
                 )
 
         # The optax rule (resolved from optax / optax.contrib), guarded by
@@ -547,6 +609,14 @@ class AbstractMultiStartGradient(AbstractMLE):
             fom_history = list(search_internal["fom_history"])
             total_steps = int(search_internal["total_steps"])
             n_resurrections = int(search_internal.get("n_resurrections", 0))
+            # ``.get`` with a default: a ``search_internal`` written before the
+            # NaN accounting existed has neither key, and a resumed run must not
+            # KeyError on it. The restored totals are lifetime counts, so a
+            # resumed run continues accumulating rather than restarting at zero.
+            n_value_nan_lane_steps = int(
+                search_internal.get("n_value_nan_lane_steps", 0)
+            )
+            n_grad_nan_lane_steps = int(search_internal.get("n_grad_nan_lane_steps", 0))
             stop_reason = search_internal.get("stop_reason", None)
 
             self.logger.info(
@@ -573,6 +643,8 @@ class AbstractMultiStartGradient(AbstractMLE):
             fom_history = []
             total_steps = 0
             n_resurrections = 0
+            n_value_nan_lane_steps = 0
+            n_grad_nan_lane_steps = 0
             stop_reason = None
 
             self.logger.info(
@@ -622,9 +694,22 @@ class AbstractMultiStartGradient(AbstractMLE):
                         self.logger.info(self._compile_message(batched=True))
                     awaiting_compile = False
 
-                foms, grads = batched_value_and_grad(params)
+                foms, grads, grad_finite = batched_value_and_grad(params)
 
-                alive = np.isfinite(np.asarray(foms))
+                # Measurement only — this accounting never gates a redraw. A
+                # gradient-NaN lane is *not* resurrected here: doing so would
+                # change search behaviour and shift every existing benchmark,
+                # so the policy question waits on what these counts show. The
+                # counters are accumulated unconditionally, outside the
+                # ``self.resurrect`` guard below, because with ``resurrect``
+                # off ``n_resurrections`` stays zero and this is then the only
+                # record that lanes died at all.
+                alive, n_value_nan, n_grad_nan = self._nan_lane_counts(
+                    foms=foms, grad_finite=grad_finite
+                )
+                n_value_nan_lane_steps += n_value_nan
+                n_grad_nan_lane_steps += n_grad_nan
+
                 foms_np = np.where(alive, np.asarray(foms), np.inf)
                 best_index = int(np.argmin(foms_np))
                 if foms_np[best_index] < best_fom:
@@ -705,6 +790,8 @@ class AbstractMultiStartGradient(AbstractMLE):
                 "fom_history": np.asarray(fom_history),
                 "total_steps": total_steps,
                 "n_resurrections": n_resurrections,
+                "n_value_nan_lane_steps": n_value_nan_lane_steps,
+                "n_grad_nan_lane_steps": n_grad_nan_lane_steps,
                 "stop_reason": stop_reason,
             }
             self.paths.save_search_internal(obj=search_internal)
@@ -942,6 +1029,17 @@ class AbstractMultiStartGradient(AbstractMLE):
             "max_consecutive_nan": self.max_consecutive_nan,
             "resurrect": self.resurrect,
             "n_resurrections": int(search_internal.get("n_resurrections", 0)),
+            # Per-step non-finite accounting, split by failure mode. Value-NaN
+            # is an undefined likelihood; gradient-NaN is a defined but
+            # non-differentiable one, counted only for lanes still alive by
+            # value so the two are disjoint. ``.get`` for the same
+            # legacy-``search_internal`` reason as ``n_resurrections``.
+            "n_value_nan_lane_steps": int(
+                search_internal.get("n_value_nan_lane_steps", 0)
+            ),
+            "n_grad_nan_lane_steps": int(
+                search_internal.get("n_grad_nan_lane_steps", 0)
+            ),
             # Auto-convergence outcome: whether the run stopped on the plateau
             # check ("converged") or exhausted the ``n_steps`` ceiling
             # ("max_steps"), the settings that produced it, and the global-best

--- a/autofit/text/text_util.py
+++ b/autofit/text/text_util.py
@@ -117,6 +117,42 @@ def search_summary_from_samples(samples) -> [str]:
     if hasattr(samples, "total_accepted_samples"):
         line.append(f"Total Accepted Samples = {samples.total_accepted_samples}\n")
         line.append(f"Acceptance Ratio = {samples.acceptance_ratio}\n")
+
+    # Per-step non-finite accounting from the gradient searches
+    # (``MultiStartGradient``). Guarded on the key rather than the search type,
+    # following the duck-typed MCMC block above: this function is on every
+    # search's summary path (``paths/abstract.py`` -> ``search_summary_to_file``),
+    # so a search whose ``samples_info`` lacks these keys — Nautilus, which is
+    # jit-only and has no gradient to be non-finite — must emit nothing at all.
+    #
+    # Rates, not just counts: raw totals are not comparable across runs, since
+    # they scale with the lane-step budget. 797 on a 16x3000 run and 10 on an
+    # 8x300 run differ 80x raw but only ~2x per lane-step.
+    #
+    # Named as the neutral facts they are. These are counts of undefined and
+    # non-differentiable likelihood evaluations; they are deliberately NOT
+    # presented as a smoothness or sampler-difficulty metric, because the
+    # correlation with (for example) HMC divergence rates is unvalidated.
+    samples_info = getattr(samples, "samples_info", None) or {}
+
+    if "n_value_nan_lane_steps" in samples_info:
+        n_value_nan = int(samples_info.get("n_value_nan_lane_steps", 0))
+        n_grad_nan = int(samples_info.get("n_grad_nan_lane_steps", 0))
+
+        line.append(f"Resurrections = {int(samples_info.get('n_resurrections', 0))}\n")
+        line.append(f"Value-NaN Lane-Steps = {n_value_nan}\n")
+        line.append(f"Gradient-NaN Lane-Steps = {n_grad_nan}\n")
+
+        # ``n_starts * total_steps`` is the number of lane-steps actually taken.
+        # A search that died before its first step has a zero denominator, so
+        # the rates are omitted rather than reported as a division error.
+        lane_steps = int(samples_info.get("n_starts", 0)) * int(
+            samples_info.get("total_steps", 0)
+        )
+        if lane_steps > 0:
+            line.append(f"Value-NaN Lane-Step Rate = {n_value_nan / lane_steps}\n")
+            line.append(f"Gradient-NaN Lane-Step Rate = {n_grad_nan / lane_steps}\n")
+
     if samples.time is not None:
         line.append(f"Time To Run = {dt.timedelta(seconds=float(samples.time))}\n")
         line.append(

--- a/test_autofit/non_linear/search/mle/test_multi_start_gradient.py
+++ b/test_autofit/non_linear/search/mle/test_multi_start_gradient.py
@@ -383,6 +383,104 @@ def test__samples_info__stop_reason_max_steps_and_legacy_search_internal():
     assert legacy.samples_info["fom_history"] is None
 
 
+@pytest.mark.parametrize(
+    "foms, grad_finite, n_value_nan, n_grad_nan",
+    [
+        # Nothing wrong: every lane finite in value and gradient.
+        ([1.0, 2.0, 3.0], [True, True, True], 0, 0),
+        # Value-NaN only -- the lane death resurrect already triggers on.
+        ([1.0, np.nan, 3.0], [True, True, True], 1, 0),
+        ([np.inf, -np.inf, 3.0], [True, True, True], 2, 0),
+        # Gradient-NaN only: likelihood defined, but not differentiable. This
+        # is the frozen-zombie lane that nothing detected before.
+        ([1.0, 2.0, 3.0], [True, False, True], 0, 1),
+        # Both failure modes on *different* lanes: counted once each.
+        ([1.0, np.nan, 3.0], [True, True, False], 1, 1),
+        # Both on the SAME lane: one value-NaN, NOT one of each. A dead lane's
+        # gradient is non-finite too, so double-counting here would inflate the
+        # gradient counter on every ordinary death and bury the real signal.
+        ([1.0, np.nan, 3.0], [True, False, True], 1, 0),
+        ([np.nan, np.nan], [False, False], 2, 0),
+    ],
+)
+def test__nan_lane_counts__splits_value_and_gradient_failures(
+    foms, grad_finite, n_value_nan, n_grad_nan
+):
+    alive, value_nan, grad_nan = af.MultiStartAdam._nan_lane_counts(
+        foms=np.asarray(foms), grad_finite=np.asarray(grad_finite)
+    )
+
+    assert value_nan == n_value_nan
+    assert grad_nan == n_grad_nan
+
+    # The alive mask is returned so the caller does not re-test finiteness.
+    assert list(alive) == list(np.isfinite(np.asarray(foms)))
+
+    # The two counts partition the lanes they describe: neither can exceed the
+    # lane count, and together they never over-count it.
+    assert value_nan + grad_nan <= len(foms)
+
+
+def test__nan_lane_counts__accepts_jax_style_array_outputs():
+    """
+    ``grad_finite`` arrives from a jitted call as a device array of bools; the
+    helper must not assume a NumPy bool dtype (a 0/1 integer array is the
+    common surprise).
+    """
+    alive, value_nan, grad_nan = af.MultiStartAdam._nan_lane_counts(
+        foms=np.asarray([1.0, 2.0, np.nan]),
+        grad_finite=np.asarray([1, 0, 1]),
+    )
+
+    assert value_nan == 1
+    assert grad_nan == 1
+    assert list(alive) == [True, True, False]
+
+
+def test__samples_info__nan_lane_step_counters():
+    model = af.Model(example.Gaussian)
+
+    best_params = np.asarray(model.vector_from_unit_vector([0.5] * model.prior_count))
+    per_start_params = np.stack([best_params])
+
+    search = af.MultiStartAdam(n_starts=1, n_steps=300)
+
+    samples = search.samples_via_internal_from(
+        model=model,
+        search_internal={
+            "params": per_start_params,
+            "best_params": best_params,
+            "best_fom": -2.0,
+            "fom_history": np.asarray([-2.0]),
+            "total_steps": 300,
+            "n_resurrections": 7,
+            "n_value_nan_lane_steps": 7,
+            "n_grad_nan_lane_steps": 3,
+            "stop_reason": "max_steps",
+        },
+    )
+
+    assert samples.samples_info["n_value_nan_lane_steps"] == 7
+    assert samples.samples_info["n_grad_nan_lane_steps"] == 3
+
+    # A search_internal written before the NaN accounting existed has neither
+    # key. Resuming or re-reading one must degrade to zero, not KeyError.
+    legacy = search.samples_via_internal_from(
+        model=model,
+        search_internal={
+            "params": per_start_params,
+            "best_params": best_params,
+            "best_fom": -2.0,
+            "fom_history": np.asarray([-2.0]),
+            "total_steps": 300,
+            "n_resurrections": 0,
+        },
+    )
+
+    assert legacy.samples_info["n_value_nan_lane_steps"] == 0
+    assert legacy.samples_info["n_grad_nan_lane_steps"] == 0
+
+
 def test__variable_length_zero_weight_nan_rows_are_robust():
     """The multi-start searches write zero-weight, NaN-log-likelihood diagnostic
     rows for the non-best starts, and auto-convergence makes runs variable-length.

--- a/test_autofit/text/test_text_util.py
+++ b/test_autofit/text/test_text_util.py
@@ -101,3 +101,115 @@ def test__search_summary_to_file(model):
     assert lines[4] == "Time Per Sample (seconds) = 0.1\n"
     assert lines[5] == "Log Likelihood Function Evaluation Time (seconds) = 1.0\n"
     results.close()
+
+
+def _gradient_samples(model, samples_info):
+    parameters = [[1.0, 2.0], [1.2, 2.2]]
+    log_likelihood_list = [1.0, 0.0]
+
+    return af.m.MockSamples(
+        model=model,
+        sample_list=af.Sample.from_lists(
+            parameter_lists=parameters,
+            log_likelihood_list=log_likelihood_list,
+            log_prior_list=[0.0, 0.0],
+            weight_list=log_likelihood_list,
+            model=model,
+        ),
+        samples_info=samples_info,
+    )
+
+
+def test__search_summary__nan_lane_step_counters(model):
+    """
+    The gradient searches report both non-finite failure modes, plus rates
+    normalised by lane-steps (``n_starts * total_steps``) -- raw counts are not
+    comparable between runs of different size.
+    """
+    samples = _gradient_samples(
+        model=model,
+        samples_info={
+            "total_samples": 10,
+            "time": None,
+            "n_starts": 8,
+            "total_steps": 100,
+            "n_resurrections": 40,
+            "n_value_nan_lane_steps": 40,
+            "n_grad_nan_lane_steps": 8,
+        },
+    )
+
+    summary = "".join(text_util.search_summary_from_samples(samples=samples))
+
+    assert "Resurrections = 40\n" in summary
+    assert "Value-NaN Lane-Steps = 40\n" in summary
+    assert "Gradient-NaN Lane-Steps = 8\n" in summary
+
+    # 40 / (8 * 100) and 8 / (8 * 100).
+    assert "Value-NaN Lane-Step Rate = 0.05\n" in summary
+    assert "Gradient-NaN Lane-Step Rate = 0.01\n" in summary
+
+
+def test__search_summary__nan_counters_absent_for_other_searches(model):
+    """
+    ``search_summary_from_samples`` is on EVERY search's summary path, so a
+    search without the counters -- Nautilus, which is jit-only and has no
+    gradient to be non-finite -- must be completely unaffected.
+    """
+    parameters = [[1.0, 2.0], [1.2, 2.2]]
+    log_likelihood_list = [1.0, 0.0]
+
+    # MockSamplesNest, not MockSamples: Nautilus produces a SamplesNest, and
+    # `total_accepted_samples` (the neighbouring duck-typed block) only exists
+    # there. The samples_info below is Nautilus's real key set -- see
+    # `NautilusSearch.samples_info_from`.
+    samples = af.m.MockSamplesNest(
+        model=model,
+        sample_list=af.Sample.from_lists(
+            parameter_lists=parameters,
+            log_likelihood_list=log_likelihood_list,
+            log_prior_list=[0.0, 0.0],
+            weight_list=log_likelihood_list,
+            model=model,
+        ),
+        samples_info={
+            "total_samples": 10,
+            "total_accepted_samples": 2,
+            "time": None,
+            "number_live_points": 1,
+            "log_evidence": 1.0,
+        },
+    )
+
+    summary = "".join(text_util.search_summary_from_samples(samples=samples))
+
+    assert "NaN" not in summary
+    assert "Resurrections" not in summary
+    assert "Lane-Step" not in summary
+
+    # The MCMC block it sits beside is untouched.
+    assert "Total Accepted Samples = 2\n" in summary
+
+
+def test__search_summary__nan_rates_omitted_when_no_lane_steps_taken(model):
+    """
+    A search that died before its first step has a zero denominator. The counts
+    still report; the rates are omitted rather than raising ZeroDivisionError.
+    """
+    samples = _gradient_samples(
+        model=model,
+        samples_info={
+            "total_samples": 10,
+            "time": None,
+            "n_starts": 8,
+            "total_steps": 0,
+            "n_resurrections": 0,
+            "n_value_nan_lane_steps": 0,
+            "n_grad_nan_lane_steps": 0,
+        },
+    )
+
+    summary = "".join(text_util.search_summary_from_samples(samples=samples))
+
+    assert "Value-NaN Lane-Steps = 0\n" in summary
+    assert "Rate" not in summary

